refactor!: Database

gavindsouza opened the pull request:
### Changes

* ~Database Pooling~
* ~Client change from `pymysql` to `mariadb`~
* Database Refactor
  * Separate Connection, Helper & Transformation layers
  * Optimize SQL execution & logging - execution of SQL faster by 30%, introduced `LazyDecode`, `LazyMogrify` methods
  * Remove dependency of non-core utils over third-party interfaces
  * Support `read_only` connections via Database class
* CI 
  * MariaDB 10.6
  * Faster fails on conflicts
  * Consolidate system dependency setup scripts
  * Skip Cypress recorder tests [don't test anything of value it seems lol] (w @ankush)
* Standardize query type checking with optimized `is_query_type` util
* Remove redundant/dead code
* Add verbosity for commands
* Better typing support on core DB APIs
* Allow `for_update` via `frappe.get_last_doc`

> Note: The client change & pooling were removed via https://github.com/frappe/frappe/pull/16961/commits/006ebcbedeb41fb21babf0a6c11a709a0a3525eb due to the added instability of MariaDB client + managing the additional system dependency. I will test and update the client once the new client reaches the complete required parity + stability. Once, we're there it's just as simple as reverting the mentioned commit to switch over :D

<!-- no-docs -->

gavindsouza committed:
refactor: Use mariadb client for creating connections

Move to use MariaDB's official Python client written in C instead of
the PyMySQL library. This change doesn't rid Frappe of the PyMySQL
library. Instead, it continues to utilize it for the ER module and
converter methods until the MariaDB library adds support for the same.

Ticket: https://jira.mariadb.org/projects/CONPY/issues/CONPY-203

gavindsouza committed:
feat(minor): Add support for `for_update` in get_last_doc

gavindsouza committed:
fix: DB error detection API usage

* Make all methods static
* Add typing hints
* Don't safe fetch attribute value for errno - MariaDB exceptions if
  raised will have errno attr in them. If the class doesn't have one,
  it's not meant to be passed in these methods.

gavindsouza committed:
refactor: Move exception detection & abstraction in separate class

This change has only been done to separate and club only like methods /
utils together

gavindsouza committed:
refactor!: frappe.db.get_database_list

* Drop mandatory unused arg in method call
* Use pluck instead of list comprehension + subscripting

gavindsouza committed:
feat: mariadb connection pooling

Start initial pool of 4 (_POOL_SIZE) connections for a given site. When
the pool is exhausted, generate new connections upon request and add
them to the pool. Max allowed size for each pool is 64 (_MAX_POOL_SIZE)
connections. However, you may have more than 64 active connections, just
that they won't be pooled but destroyed upon job completion/end of
request/similar cycle.

gavindsouza committed:
fix(db)!: Base methods should throw NotImplementedError

gavindsouza committed:
fix: frappe.conf.disable_database_connection_pooling

For the times you don't want to use pooling ;) Calling
get_connection/connect will close all pooled connections
for the current process / worker!

gavindsouza committed:
fix: Separate pool for replica setup

_SITE_POOLS[frappe.local.site].read_only will hold ConnectionPool for
replica database connections. _SITE_POOLS[frappe.local.site].default
will hold pool for connections that allow read+writes

gavindsouza committed:
refactor: Move connection & pool management out, inherit instead

gavindsouza committed:
fix!: Remove frappe.db.create_help_table

Help table has been deprecated for a while. It's safe to get rid of the
API too.

gavindsouza committed:
fix: Depend on replica details being there for Replica ConnectionPool

This logic mirror how replica connections are handled

gavindsouza committed:
ci: Install system dependency - libmariadb-dev

gavindsouza committed:
fix: F821 undefined name 'InternalError'

tx flake8

gavindsouza commented:
If you get an error while setting this up that looks like:

```python
Collecting mariadb
  Using cached https://files.pythonhosted.org/packages/8f/c9/7050899dc1066409a17e1147d3afe1b078e582afdb755c6d3cb9c9a5c3ab/mariadb-1.0.0.tar.gz
    Complete output from command python setup.py egg_info:
    /bin/sh: 1: mariadb_config: not found
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-2gdw_t_r/mariadb/setup.py", line 26, in <module>
        cfg = get_config(options)
      File "/tmp/pip-build-2gdw_t_r/mariadb/mariadb_posix.py", line 49, in get_config
        cc_version = mariadb_config(config_prg, "cc_version")
      File "/tmp/pip-build-2gdw_t_r/mariadb/mariadb_posix.py", line 27, in mariadb_config
        "mariadb_config not found.\nPlease make sure, that MariaDB Connector/C is installed on your system, edit the configuration file 'site.cfg' and set the 'mariadb_config'\noption, which should point to the mariadb_config utility.")
    OSError: mariadb_config not found.
    Please make sure, that MariaDB Connector/C is installed on your system, edit the configuration file 'site.cfg' and set the 'mariadb_config'
    option, which should point to the mariadb_config utility.
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-2gdw_t_r/mariadb/
```

You need to install the MariaDB database development files package (or `libmariadb-dev`). Refer to https://stackoverflow.com/a/63326550/10309266

gavindsouza committed:
fix: Allow non-blocking checks via MariaDBExceptionUtils

gavindsouza committed:
refactor(minor): Move internal util to the module where it's used

gavindsouza committed:
fix: frappe.DISABLE_DATABASE_POOLING to override frappe.conf.disable_database_connection_pooling

Also, don't pool root connections

gavindsouza committed:
fix(drop-site): Tell user where archived site data is stored

gavindsouza committed:
refactor: DBManager

* Simplify logic: DRY, lesser indentation & all DAT
* Utilize newer APIs, f-strings & more
* Cleaner namespace
* Conform inconsistent behaviours

gavindsouza committed:
fix: Remove half baked support for filters_config in db.query

gavindsouza committed:
fix: Connection Pooling

* Reduce _POOL_SIZE from 4 to 1. New pools will have just one
  connection. They can scale up as per requirement there after.
* Set auto_connect flag in MariaDB connection - https://mariadb.com/docs/connect/programming-languages/python/connect/

gavindsouza committed:
fix: Correct use of is_connection_pooling_enabled check

gavindsouza committed:
fix: Use sql_ddl instead of sql

Previous attempts to run this would lead to stalls...long long stalls.
And nothing blocking the query, it would stay in a "Waiting for table
metadata lock" state.

gavindsouza committed:
fix: Disable db pooling on all commands by default

Set envvar DATABASE_POOLING to enable

gavindsouza committed:
refactor: Base Database class

* DRY, explicit > implicit usages
* Don't re-compute and do multiple calls for errprint, log, mogrify, etc
* Use consistent logging methods
* Simplify logic / Use newer APIs where applicable

gavindsouza committed:
refactor: Database

* Change query notations - QB > raw
* Update logic of DB APIs - simplify & perf improvements

gavindsouza committed:
feat(db): _transform_query to convert args passed to db.cursor

Transform query & parameters based on client requirements. Eg: MariaDB
client doesn't behave similar to PyMySQL or Psycopg2.

ref: https://jira.mariadb.org/projects/CONPY/issues/CONPY-205

gavindsouza committed:
refactor(minor): PostgresDatabase's exceptions

Focus on Non-blocking Postgres exception checking by:

* Use safe getattr to fetch pgcode instead
* Use psycopg errorcodes module to use named variables instead of direct
  codes...for way superior readability xD

Also, moved exceptions out of the main class - just code separation, no
namespace change.

gavindsouza commented:
MariaDB Client 1.1.0 will be released this month, we should probably wait for that release before completing this 🤔

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into mariadb-client-refactor

gavindsouza committed:
fix(db): Log queried tables through generated query

gavindsouza committed:
fix(db): Skip transformation of query if no named params found

gavindsouza committed:
fix(qb): Use fallback og table if not found in schema mapper

gavindsouza committed:
fix(db): Track query engine under db._filter_engine

gavindsouza committed:
refactor: frappe.db.sql

* Move everything except _cursor.execute outside try-except block - This
  caused multiple traceback printing (by recursion of db.sql)
* Include values mogrifying & executing via client alone in execution
  time for query
* Reduce indentations lol

gavindsouza committed:
refactor: frappe.db.log_query

* Mogrify queries and set them as frappe.db.last_query instead of
  directly interfacing with the clients
* This is required for now as the MariaDB client uses binary protocol to
  talk to the server and doesn't build the queries itself
* Add typing hints
* Imported Query object as FilterEngine - Query is too ambiguous lol

ref: https://jira.mariadb.org/browse/CONPY-208?focusedCommentId=226873&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-226873

gavindsouza committed:
fix: Recorder to use frappe.db.last_query to pick out last executed stmt

gavindsouza committed:
perf: Use lazy mogrified query for logging

For parameterized queries, there's an improvement of ~30% in query
execution via frappe.db.sql - from 58.6 µs ± 2.37 µs to 44.6 µs ± 1.56 µs

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
chore: NoneType alias for < PY310

gavindsouza committed:
fix(db): Store result of last executed query under frappe.db.last_result

gavindsouza committed:
test: Get rid of magical forgiving behaviour of pymysql

("test",) would be interpretted as ("test") which would be just "test".
MariaDB client doesn't handle errenous inputs like this that PyMySQL
tolerated

gavindsouza committed:
test: frappe.db.describe returns List[Tuple] through mariadb

gavindsouza committed:
fix(db): Transform values only if not None

gavindsouza committed:
test(sequence): Use mariadb client's exception handling

gavindsouza committed:
fix(sequence): Setup & use SequenceGeneratorLimitExceeded error

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
fix: Add fallbacks for values

Psycopg seems to like None over () and MariaDB - PyMySQL can't seem to
agree on anything - so this seems to keep everyone happy...a very
delicate balance :crie:

gavindsouza committed:
refactor(minor): Use db.db_type instead of conf.db_type

gavindsouza committed:
fix: Make postgres' last_query lazy decodable

gavindsouza committed:
fix(get_contact_list): Don't pass conditions as query value

* Conditions passed are not valid prepared statement values. They
can be passed as string substitution since they're generated by DBQuery.
* Added typing hints & other improvements
* Removed seemingly pointless try-except block

gavindsouza committed:
fix: Improve _transform_query unique key identification

gavindsouza committed:
refactor: DISABLE_DATABASE_CONNECTION_POOLING conf + var name

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
Revert "fix: Disable db pooling on all commands by default"

This reverts commit 0a8941c58317956e0661f90fa2357e39674f63c3 since it
may not be required as we start pooling with single connections only.

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
refactor(db-read_only): Track conn type in Database instance

gavindsouza committed:
fix: Disable connection pooling via bench commands unless specified

gavindsouza committed:
refactor(run-ui-tests): Maintain list of cypress plugins

gavindsouza commented:
### Pending Actions

- [x] Load testing on production setup - https://conn-pool-000.frappe.cloud/
- [x] Pass failing Cypress tests
- [ ] Run ERPNext CI against branch - https://github.com/frappe/erpnext/pull/31454

gavindsouza committed:
ci(ui-tests): Print 'bench start' log on failure

Co-authored-by: Ankush Menat <ankush@frappe.io>

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

ankush committed:
chore: disable recorder UI test

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza commented:
Cypress is going crazy taking different versions of Chrome on different parallel builds 😢. We'd probs have to update from `6.9.1` to the later versions (like `10.2.0`) for setting the fixes, etc - but the [migration guide](https://docs.cypress.io/guides/references/migration-guide) is huge and we'd rather take it on in a separate PR 🥲

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza commented:
MariaDB client doesn't take tuple/list args for `"name" in List` queries (ref: `mariadb.DataError: Data type 'tuple' in column 0 not supported in MariaDB Connector/Python` via https://github.com/frappe/erpnext/runs/7089952460?check_suite_focus=true) - will have to add a workaround for that in `frappe.db.sql` ig 🤔 - although I'm leaning towards just making peace with db.sql not supporting that.

Edit: Handled using pymysql's `escape_sequence`

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into mariadb-client-refactor

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
fix(db): Hanlde sequences in db.sql values

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into mariadb-client-refactor

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into mariadb-client-refactor

gavindsouza committed:
fix: Database._transform_query

* Handle dict substitutions in transformations too
* Allow list / tuple values
* Check for values inconsistencies before flattening step

gavindsouza committed:
chore(deps): Bump MariaDB client from 1.0.11 to 1.1.2

gavindsouza committed:
ci: Separate PY / conflict validation step, merge install

gavindsouza commented:
Will resolve conflicts after merging #15891 which is likely going to add more 😄

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into mariadb-client-refactor

gavindsouza committed:
ci: Add mariadb system dependency

gavindsouza committed:
fix: Transform queries with all types of `values`

gavindsouza committed:
fix: Use mariadb's constants instead of pymysql

gavindsouza committed:
fix: Format string instead of % to avoid TypeError

gavindsouza committed:
fix(pg): Transform Falsy values as None

ref: https://github.com/frappe/frappe/runs/7208763442

gavindsouza committed:
fix: Database._transform_result

Transform data not already casted by mariadb client

gavindsouza committed:
fix(mariadb): SequenceGeneratorLimitExceeded is an OperationalError now
:D

gavindsouza committed:
fix: Database._transform_query

* Don't try to convert None to sequence
* Skip transforming values that mariadb~=1.1 can handle - only worry
  about sequences for now

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into mariadb-client-refactor

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
refactor: Use pymysql over mariadb client

This is supposed to be a temporary switch to make the parent PR easier
to digest. MariaDB client has some issues with release, and system
dependencies.

This commit may be reverted to enable mariadb client again.

gavindsouza committed:
test: More resilient tests for sequences

gavindsouza commented:
ERPNext CI: https://github.com/frappe/erpnext/actions/runs/2711298269

gavindsouza committed:
fix(db): Revert breaking change of wrapping NoneType in sequence

gavindsouza committed:
fix(db): Use sentinel object for default values paramters

ankush commented on frappe/__init__.py:
```diff
@@ -48,6 +48,7 @@
 controllers = {}
 local = Local()
 STANDARD_USERS = ("Guest", "Administrator")
+DISABLE_DATABASE_CONNECTION_POOLING = None
```
this should be in that revert PR? 👀

ankush replied:
This and its references should be part of that revert commit?

ankush commented on frappe/__init__.py:
```diff
@@ -283,7 +284,9 @@ def connect_replica():
 
-	local.replica_db = get_db(host=local.conf.replica_host, user=user, password=password, port=port)
+	local.replica_db = get_db(
+		host=local.conf.replica_host, user=user, password=password, port=port, read_only=True
```
Why read-only by default?

ankush commented on frappe/__init__.py:
```diff
@@ -1180,11 +1183,11 @@ def get_doc(*args, **kwargs) -> "Document":
 
 
-def get_last_doc(doctype, filters=None, order_by="creation desc"):
+def get_last_doc(doctype, filters=None, order_by="creation desc", for_update=False):
```
```suggestion
def get_last_doc(doctype, filters=None, order_by="creation desc", *, for_update=False):
```

ankush commented on frappe/__init__.py:
```diff
@@ -283,7 +284,9 @@ def connect_replica():
 
-	local.replica_db = get_db(host=local.conf.replica_host, user=user, password=password, port=port)
+	local.replica_db = get_db(
+		host=local.conf.replica_host, user=user, password=password, port=port, read_only=True
```
read only by default? 👀

gavindsouza replied:
Because replica's are used for read only actions? Looking at the `@read_only` wrapper

ankush commented on frappe/__init__.py:
```diff
@@ -1180,11 +1183,11 @@ def get_doc(*args, **kwargs) -> "Document":
 
 
-def get_last_doc(doctype, filters=None, order_by="creation desc"):
+def get_last_doc(doctype, filters=None, order_by="creation desc", for_update=False):
```
```suggestion
def get_last_doc(doctype, filters=None, order_by="creation desc", *, for_update=False):
```

Future proofing.

gavindsouza replied:
I would wish to enforce this after the first two params - and for other APIs too. Should raise a separate PR for that tho

ankush commented on frappe/database/database.py:
```diff
@@ -73,6 +79,8 @@ def __init__(self, host=None, user=None, password=None, ac_name=None, use_defaul
 
 		self.password = password or frappe.conf.db_password
 		self.value_cache = {}
+		# self.db_type: str
+		# self.last_query (lazy) attribute of last sql query executed
```
```suggestion
```

gavindsouza committed:
Merge branch 'develop' into mariadb-client-refactor

gavindsouza committed:
fix: Set default port attribute for Database classes

db.default_port wil be available as a class attribute to hold defaults
for DB types.

Usage: frappe.conf.db_port or frappe.db.default_port
Why: I couldn't run the mariadb command because the defaults aren't set
for my system. server is remote / containerized. Setting port in
equivalent mysql command fixes this.

gavindsouza committed:
fix(db*): Use common fallback Datetime str

There existed inconsistencies between db_query & db's fallback for min
datetime in str format - missing decimal seconds places. Now, we're
storing the default string once and re-using it to reduce
inconsistencies or room for human errors.

gavindsouza committed:
chore: Deprecate backup script + minor refactors

gavindsouza committed:
test(db): Add tests for untested db methods

gavindsouza committed:
fix(get_last_doc): Allow for_update as kwarg only

Co-authored-by: Ankush Menat <ankushmenat@gmail.com>

gavindsouza commented:
The last ERPNext CI build was all green, re-triggered it for confirmation.

gavindsouza commented:
All Passing 🥳 for Frappe & ERPNext

![image](https://user-images.githubusercontent.com/36654812/180401685-0df131d3-efc8-42c2-af56-6f192c55e7b4.png)
![image](https://user-images.githubusercontent.com/36654812/180401753-8b2dd150-3edf-497f-a38f-2642f4227787.png)
